Getting failure while retrieving feature flag on Android/kotlin Client

result: Failure(java.lang.NullPointerException
                                                                                                    	com.github.kittinunf.fuel.core.FuelError$a.a(FuelError.kt:39)
                                                                                                    	com.github.kittinunf.fuel.core.FuelError$a.b(FuelError.kt:13)
                                                                                                    	t2.g.call(RequestTaskCallbacks.kt:53)
                                                                                                    Caused by: java.lang.NullPointerException
                                                                                                    	m2.b.o(Flagsmith.kt:90)
                                                                                                    	m2.a.k(Flagsmith.kt:39)
                                                                                                    	q2.g$a.k(Deserializable.kt:27)
                                                                                                    	q2.h.p(Deserializable.kt:19)
                                                                                                    	q2.y.run(Unknown Source:2)
                                                                                                    	q2.e$a.execute(Environment.kt:3)
                                                                                                    	q2.j.o(Deserializable.kt:52)
                                                                                                    	t2.g.call(RequestTaskCallbacks.kt:35)
                                                                                                    	java.util.concurrent.FutureTask.run(FutureTask.java:264)
                                                                                                    	java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1145)
                                                                                                    	java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:644)
                                                                                                    	java.lang.Thread.run(Thread.java:1012)
                                                                                                    )

Hi @dabeeeenster while using v1.0.4 , I am getting below attached error when we initialized flag smit object:



`FATAL EXCEPTION: main
                                                                                                    Process: one.upswing.edhas, PID: 15379
                                                                                                    java.lang.RuntimeException: Unable to start activity ComponentInfo{one.upswing.edhas/one.upswing.edhas.ui.ExploreSdkActivity}: kotlin.UninitializedPropertyAccessException: lateinit property retrofit has not been initialized
                                                                                                    	at android.app.ActivityThread.performLaunchActivity(ActivityThread.java:3967)
                                                                                                    	at android.app.ActivityThread.handleLaunchActivity(ActivityThread.java:4112)
                                                                                                    	at android.app.servertransaction.LaunchActivityItem.execute(LaunchActivityItem.java:101)
                                                                                                    	at android.app.servertransaction.TransactionExecutor.executeCallbacks(TransactionExecutor.java:135)
                                                                                                    	at android.app.servertransaction.TransactionExecutor.execute(TransactionExecutor.java:95)
                                                                                                    	at android.app.ActivityThread$H.handleMessage(ActivityThread.java:2527)
                                                                                                    	at android.os.Handler.dispatchMessage(Handler.java:106)
                                                                                                    	at android.os.Looper.loopOnce(Looper.java:223)
                                                                                                    	at android.os.Looper.loop(Looper.java:324)
                                                                                                    	at android.app.ActivityThread.main(ActivityThread.java:8485)
                                                                                                    	at java.lang.reflect.Method.invoke(Native Method)
                                                                                                    	at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:582)
                                                                                                    	at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:1061)
                                                                                                    Caused by: kotlin.UninitializedPropertyAccessException: lateinit property retrofit has not been initialized
                                                                                                    	at m2.b.<init>(Flagsmith.kt:145)
                                                                                                    	at one.upswing.edhas.ui.ExploreSdkActivity.onCreate(ExploreSdkActivity.kt:78)
                                                                                                    	at android.app.Activity.performCreate(Activity.java:8733)
                                                                                                    	at android.app.Activity.performCreate(Activity.java:8702)
                                                                                                    	at android.app.Instrumentation.callActivityOnCreate(Instrumentation.java:1384)
                                                                                                    	at android.app.ActivityThread.performLaunchActivity(ActivityThread.java:3948)
                                                                                                    	at android.app.ActivityThread.handleLaunchActivity(ActivityThread.java:4112) 
                                                                                                    	at android.app.servertransaction.LaunchActivityItem.execute(LaunchActivityItem.java:101) 
                                                                                                    	at android.app.servertransaction.TransactionExecutor.executeCallbacks(TransactionExecutor.java:135) 
                                                                                                    	at android.app.servertransaction.TransactionExecutor.execute(TransactionExecutor.java:95) 
                                                                                                    	at android.app.ActivityThread$H.handleMessage(ActivityThread.java:2527) 
                                                                                                    	at android.os.Handler.dispatchMessage(Handler.java:106) 
                                                                                                    	at android.os.Looper.loopOnce(Looper.java:223) 
                                                                                                    	at android.os.Looper.loop(Looper.java:324) 
                                                                                                    	at android.app.ActivityThread.main(ActivityThread.java:8485) 
                                                                                                    	at java.lang.reflect.Method.invoke(Native Method) 
                                                                                                    	at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:582) 
                                                                                                    	at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:1061)`
                                                                                                    	
                                                                                                    	
                                                                                                    	
Below are the steps I am using to integrate flagsmit:

app's builld.gralde dependency
`implementation 'com.github.Flagsmith:flagsmith-kotlin-android-client:1.4.0'`
Project level build.gradle
`maven { url "https://jitpack.io" }`

in Activity's onCreate function
```
val flagsmith = Flagsmith(environmentKey = "Uoo5YF76zoWec9zyigDZkH", context = this, enableAnalytics = true)
  flagsmith.getValueForFeature("td_amount") { result ->
                    Log.d("Flagsmith", "result: $result")
                    Log.d("Flagsmith", "result: ${result.isFailure}")
                    Toast.makeText(this, "TD Amount: $result", Toast.LENGTH_LONG).show()
                }

```

--- a/FlagsmithClient/src/main/java/com/flagsmith/Flagsmith.kt
+++ b/FlagsmithClient/src/main/java/com/flagsmith/Flagsmith.kt
@@ -45,10 +45,7 @@ class Flagsmith constructor(
     private lateinit var retrofit: FlagsmithRetrofitService
     private var cache: Cache? = null
     private var lastUsedIdentity: String? = null
-    private val analytics: FlagsmithAnalytics? =
-        if (!enableAnalytics) null
-        else if (context != null) FlagsmithAnalytics(context, retrofit, analyticsFlushPeriod)
-        else throw IllegalArgumentException("Flagsmith requires a context to use the analytics feature")
+    private var analytics: FlagsmithAnalytics? = null //TODO: Make this a lateinit var and instead initialise in init block
 
     private val eventService: FlagsmithEventService? =
         if (!enableRealtimeUpdates) null
@@ -93,6 +90,13 @@ class Flagsmith constructor(
             writeTimeoutSeconds = writeTimeoutSeconds, timeTracker = this, klass = FlagsmithRetrofitService::class.java)
         retrofit = pair.first
         cache = pair.second
+
+        if (enableAnalytics) {
+            if (context == null || context.applicationContext == null) {
+                throw IllegalArgumentException("Flagsmith requires a context to use the analytics feature")
+            }
+            analytics = FlagsmithAnalytics(context, retrofit, analyticsFlushPeriod)
+        }
     }
 
     companion object {

--- a/FlagsmithClient/src/main/java/com/flagsmith/Flagsmith.kt
+++ b/FlagsmithClient/src/main/java/com/flagsmith/Flagsmith.kt
@@ -45,7 +45,7 @@ class Flagsmith constructor(
     private lateinit var retrofit: FlagsmithRetrofitService
     private var cache: Cache? = null
     private var lastUsedIdentity: String? = null
-    private var analytics: FlagsmithAnalytics? = null //TODO: Make this a lateinit var and instead initialise in init block
+    private var analytics: FlagsmithAnalytics? = null
 
     private val eventService: FlagsmithEventService? =
         if (!enableRealtimeUpdates) null

--- a/FlagsmithClient/src/test/java/com/flagsmith/FeatureFlagCachingTests.kt
+++ b/FlagsmithClient/src/test/java/com/flagsmith/FeatureFlagCachingTests.kt
@@ -77,7 +77,7 @@ class FeatureFlagCachingTests {
         flagsmithWithCache = Flagsmith(
             environmentKey = "",
             baseUrl = "http://localhost:${mockServer.localPort}",
-            enableAnalytics = false,
+            enableAnalytics = true, // Mix up the analytics flag to test initialisation
             context = mockApplicationContext,
             defaultFlags = defaultFlags,
             cacheConfig = FlagsmithCacheConfig(enableCache = true)
@@ -112,6 +112,7 @@ class FeatureFlagCachingTests {
         `when`(mockContextResources.getBoolean(anyInt())).thenReturn(false)
         `when`(mockContextResources.getDimension(anyInt())).thenReturn(100f)
         `when`(mockContextResources.getIntArray(anyInt())).thenReturn(intArrayOf(1, 2, 3))
+        `when`(mockApplicationContext.applicationContext).thenReturn(mockApplicationContext)
     }
 
     @After

